proposals/nnnn-safe-shared-contiguous-storage.md

```diff
@@ -0,0 +1,1179 @@
+
+Even if `Span` were to replace all uses of a constant `Data` in API, something like `Data` would still be needed, just as `Array<T>` will: resizing mutations (e.g. `RangeReplaceableCollection` conformance.) We may still want to add an untyped-element equivalent of `Array` at a later time.
+
+##### <a name="Conversions"></a>Syntactic Sugar for Automatic Conversions
+In the context of a resilient library, a generic entry point in terms of `some ContiguousStorage` may add unwanted overhead. As detailed above, an entry point in an evolution-enabled library requires an inlinable  generic public entry point which forwards to a publicly-accessible function defined in terms of `Span`. If `Span` does become a widely-used type to interface between libraries, we could simplify these conversions with a bit of compiler help.
+
+We could provide an automatic way to use a `ContiguousStorage`-conforming type with a function that takes a `Span` of the appropriate element type:
+
+```swift
+func myStrnlen(_ b: Span<UInt8>) -> Int {
+  guard let i = b.firstIndex(of: 0) else { return b.count }
+  return b.distance(from: b.startIndex, to: e)
+}
+let data = Data((0..<9).reversed()) // Data conforms to ContiguousStorage
+let array = Array(data) // Array<UInt8> also conforms to ContiguousStorage
+myStrnlen(data)  // 8
+myStrnlen(array) // 8
+```
+
+This would probably consist of a new type of custom conversion in the language. A type author would provide a way to convert from their type to an owned `Span`, and the compiler would insert that conversion where needed. This would enhance readability and reduce boilerplate.
+
+##### Interopability with C++'s `std::span` and with llvm's `-fbounds-safety`
+The [`std::span`](https://en.cppreference.com/w/cpp/container/span) class template from the C++ standard library is a similar representation of a contiguous range of memory. LLVM may soon have a [bounds-checking mode](https://discourse.llvm.org/t/70854) for C. These are opportunities for better, safer interoperation with Swift, via a type such as `Span`.
+
+## Acknowledgments
+
+Joe Groff, John McCall, Tim Kientzle, Karoy Lorentey contributed to this proposal with their clarifying questions and discussions.
+
+
+
+## <a name="Indexing"></a>Appendix: Index and slicing design considerations
+
+Early prototypes of this proposal defined an `Index` type, `Iterator` types, etc. We are proposing `Int`-based API and are deferring defining `Index` and `Iterator` until more of the non-escapable collection story is sorted out. The below is some of our research into different potential designs of an `Index` type.
```

Comment:
We are okay with having an `Index` type on `Span`, and we must be proposing it above. (`UnsafeBufferPointer` has sunk the concrete concept of an `Index` down into the ownership model in SE-0437. Whatever shape we will propose for a collection-like container of noncopyable elements, it will include the idea of an escaping `Index`, with this exact name.)

(`Iterator` on the other hand is not the right name for a borrowing iterator type, as the classic `Sequence` protocol reserves this name for its unified borrowing+consuming iterator. That iterator shape does not translate to the ownership model. `Sequence` will be split into two+ variants, with each variant coming with a distinct name for its iterators. `Span` will soon need to define a borrowing iterator type; however, this work must be deferred until we have introduced that concept -- the addition should be done in that same proposal.)